borsh-derive/src/internals/schema/structs/mod.rs

```diff
@@ -62,10 +62,7 @@ pub fn process(input: &ItemStruct, cratename: Path) -> syn::Result<TokenStream2>
         fn add_definitions_recursively(definitions: &mut #cratename::__private::maybestd::collections::BTreeMap<#cratename::schema::Declaration, #cratename::schema::Definition>) {
             #struct_fields
             let definition = #cratename::schema::Definition::Struct { fields };
-
-            let no_recursion_flag = definitions.get(&<Self as #cratename::BorshSchema>::declaration()).is_none();
-            #cratename::schema::add_definition(<Self as #cratename::BorshSchema>::declaration(), definition, definitions);
-            if no_recursion_flag {
```

Comment:
In my opinion, there's no point in propagating this single loophole, needed for derives of `BorshSchema` for recursive structs/enums, everywhere, and, changing the signature of `borsh::schema::add_definition` as a convenience method for that purpose, 
because it decreases the number of cases where schema `Definition` collisions are detected (2 `Definitions` mapped to the same `Declaration`).
Unfortunately, collissions illustrated by `test_implicit_conflict_struct` and `test_implicit_conflict_enum` aren't detected on current master either, but a more robust solution would probably require more significant changes to `BorshSchema` trait and much more complex code overall. 
Let's keep it as it is now.